src/storage-controller/src/collection_mgmt.rs

```diff
@@ -192,13 +346,377 @@ where
+        || format!("CollectionManager-incremental_write_task-{id}"),
+        async move {
+            let mut interval = tokio::time::interval(Duration::from_millis(DEFAULT_TICK_MS));
+
+            // This is memory inefficient: we always keep a full copy of
+            // desired, so that we can re-derive a to_write if/when someone else
+            // writes to persist and we notice because of an upper conflict.
+            // This is optimized for the case where we rarely have more than one
+            // writer.
+            //
+            // We can optimize for a multi-writer case by keeping an open
+            // ReadHandle and continually reading updates from persist, updating
+            // a desired in place. Similar to the self-correcting persist_sink.
+            let mut desired = Vec::new();
+            let mut to_write = Vec::new();
+
+            let mut current_upper = {
+                let upper_result = match write_handle.recent_upper(id.clone()).await {
+                    // We got a response!
+                    Ok(upper) => upper,
+                    // Failed to receive which means the worker shutdown.
+                    Err(_recv_error) => {
+                        // End the task since we can no longer send writes to persist.
+                        return;
+                    }
+                };
+
+                let upper = match upper_result {
+                    Ok(upper) => upper,
+                    Err(other) => {
+                        panic!("Unhandled error while fetching upper of incremental managed collection {id:?}: {other:?}")
+                    }
+                };
+
+                if upper.is_empty() {
+                    return;
+                }
+
+                upper.into_option().expect("known to exist")
+            };
+
+            // `desired` starts out empty, so we initialize `to_write` to be a
+            // retraction of all that's currently in persist.
+            {
+                if current_upper == T::minimum() {
+                    // Nothing there, so far!
+                } else {
+                    let mut read_handle = read_handle_fn().await;
+                    let as_of = current_upper.step_back().unwrap_or_else(|| T::minimum());
+                    let as_of = Antichain::from_elem(as_of);
+                    let snapshot = read_handle.snapshot_and_fetch(as_of).await;
+
+                    let mut negated_oks = match snapshot {
+                        Ok(contents) => {
+                            let mut snapshot = Vec::with_capacity(contents.len());
+                            for ((data, _), _, diff) in contents {
+                                let row = data.expect("invalid protobuf data").0.unwrap();
+                                snapshot.push((row, -diff));
+                            }
+                            snapshot
+                        }
+                        Err(_) => panic!("read before since"),
```

Comment:
This seems like something that can happen if a storage collection is created with a since != `[0]`. I think the storage controller will only bump the since on creation, not the upper. Can we handle this gracefully by first advancing the upper to the since?